### DISCUSSION TITLE
Should we change the type of the time variables?

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/evm-monorepo#122] [General]
In the actual implementation of the `core` contracts we are storing in the `Stream` struct only `uint256` type variables, which are expensive in terms of gas cost. 
https://github.com/sablierhq/v2-core/blob/c7cbed688b87facd042dfeaa869882b7a3c66dc5/src/interfaces/ISablierV2Linear.sol#L56-L66
https://github.com/sablierhq/v2-core/blob/c7cbed688b87facd042dfeaa869882b7a3c66dc5/src/interfaces/ISablierV2Pro.sol#L81-L93

I was toying with all the `uint` types that solidity has and I think the best solution is to change the `uint256` type to `uint64`  for  our time variables: `cliffTime` `segmentMilestones` `startTime` and `stopTime`.

The only assumption we have to make is that all timestamps can fit in `uint64`, i.e. this number 18446744073709551615, this will be true until the year ~584942417303, means it is totally safe.

I've actually created a branch 0eeba8cd2e3761088fb76d88bd5a1f23a6ee1382 to test the gas benefits and I was surprised to find out that this change would save us **43,212** gas in average for the `create` function in the `SablierV2Linear` contract and **140,236** in the `SablierV2Pro` contract.

Here are some of the screenshots of the gas reports:
## Before
From this commit: 3003b344749550c9bd07f4840af4c737c6f157f8
https://imgur.com/a/fooGyFr

## After
From this commit: 0eeba8cd2e3761088fb76d88bd5a1f23a6ee1382
https://imgur.com/a/HoIBp0a

**Comment by PaulRBerg** [comment]
Interesting proposal! I'm tentatively in favor of implementing this - let me just double check that this doesn't introduce any bug or breaking change and that I can reproduce the gas savings you reported.

> i.e. this number 18446744073709551615, this will be true until the year ~584942417303, means it is totally safe

LOL. Unix cannot even go that far. Read about the [Year 2038 problem](https://en.wikipedia.org/wiki/Year_2038_problem).

>  I was surprised to find out that this change would save us 43,212 gas on average for the `create` function in the `SablierV2Linear` contract and 140,236 in the `SablierV2Pro` contract.

The gas savings are high because storing anything in contract storage is expensive.

Though I'm not sure that they are as high as you claim them to be - to get a more accurate picture, we might have to compare only the cases when the creation of the stream succeeds, because that's when the struct is created and the time variables are actually saved in storage.

**Reply by andreivladbrg** [reply]
> Interesting proposal! I'm tentatively in favor of implementing this - let me just double check that this doesn't introduce any bug or breaking change and that I can reproduce the gas savings you reported.

Glad you like it.

> The gas savings are high because storing anything in contract storage is expensive.

I know, that's correct.

> Though I'm not sure that they are high as you claim them to be - to get a more accurate picture, we might have to compare only the cases when the creation of the stream succeeds, because that's when the struct is created and the time variables are actually saved in storage.

Actually, I was expecting higher numbers considering the way I packed the variables in the struct. I saved 3 blocks of 256 bits from the storage, that is equivalent to ~60,000 gas.

Here are some screenshots of the gas reports with only pass tests.

<Details>

## Before
<img width="871" alt="Screen Shot 2022-08-24 at 00 32 45" src="https://user-images.githubusercontent.com/99738872/186270870-ee394c5c-a08c-459e-b437-d770a796494c.png">

## After
<img width="870" alt="Screen Shot 2022-08-24 at 00 32 35" src="https://user-images.githubusercontent.com/99738872/186270897-93c1ed7d-4cb9-4e5a-a83b-5d6d7188ea1c.png">

</Details>

**Reply by PaulRBerg** [reply]
Great, thanks for doing a gas comparison for only the case when the stream creation succeeds.

> I saved 3 blocks of 256 bits from the storage, that is equivalent to ~60,000 gas.

Yes, but you didn't order them in the optimal way. For example, in the linear [`Stream`](https://github.com/sablierhq/v2-core/blob/0eeba8cd2e3761088fb76d88bd5a1f23a6ee1382/src/interfaces/ISablierV2Linear.sol#L56-L65) struct, you should have defined the `cliffTime` after the `token`, because the latter is encoded as `uint160`, and the optimal way to order fields in a struct is to place the smallest bit-width variables at the bottom. So, ideally we would put the `cliffTime` after `startTime`.

**Reply by andreivladbrg** [reply]
> Yes, but you didn't order them in the optimal way.

Well, the most important thing is to have only 5 blocks of storage, what you suggest would bring only a small change, but it is welcome.

**Reply by PaulRBerg** [reply]
It would actually be a significant change, since Solidity would pack the 4 variables together in the same EVM word.

Now it only packs the last 3 variables.

Try it and see what results you get!

**Reply by andreivladbrg** [reply]
> Now it only packs the last 3 variables.

It packs only the last 2 variables.

I think you are missing the fact that the `sender` is packed together with `cliffTime` and `token` with `startTime` because 160+64<256(the block size).

I actually tested and there is a sligthly **increase** of 9 gas on your suggestion.

**Reply by PaulRBerg** [reply]
You're right, I did not realize that. My suggestion is more expensive.

May I at least suggestion we swap our the places for `cliffTime` and `startTime`?

So `startTime` to appear after `sender`, and `cliffTime` after `token`, like this:

```solidity
struct Stream {
    uint256 depositAmount;
    uint256 withdrawnAmount;
    address sender;
    uint64 cliffTime;
    IERC20 token;
    uint64 startTime;
    uint64 stopTime;
    bool cancelable;
}
```

**Comment by PaulRBerg** [comment]
This has been addressed by #132. Locking.
